src/Npgsql/Internal/NpgsqlConnector.FrontendMessages.cs

```diff
@@ -201,30 +199,38 @@ internal async Task WriteParse(string sql, string statementName, List<NpgsqlPara
         // 0 length implicitly means all-text, 1 means all-binary, >1 means mix-and-match
         if (formatCodeListLength == 1)
         {
-            if (WriteBuffer.WriteSpaceLeft < 2)
+            if (WriteBuffer.WriteSpaceLeft < sizeof(short))
                 await Flush(async, cancellationToken).ConfigureAwait(false);
-            WriteBuffer.WriteInt16((short)FormatCode.Binary);
+            WriteBuffer.WriteInt16(DataFormat.Binary.ToFormatCode());
         }
         else if (formatCodeListLength > 1)
         {
             for (var paramIndex = 0; paramIndex < parameters.Count; paramIndex++)
             {
                 if (WriteBuffer.WriteSpaceLeft < 2)
                     await Flush(async, cancellationToken).ConfigureAwait(false);
-                WriteBuffer.WriteInt16((short)parameters[paramIndex].FormatCode);
+                WriteBuffer.WriteInt16(parameters[paramIndex].Format.ToFormatCode());
             }
         }
 
-        if (WriteBuffer.WriteSpaceLeft < 2)
+        if (WriteBuffer.WriteSpaceLeft < sizeof(ushort))
             await Flush(async, cancellationToken).ConfigureAwait(false);
 
         WriteBuffer.WriteUInt16((ushort)parameters.Count);
 
-        for (var paramIndex = 0; paramIndex < parameters.Count; paramIndex++)
+        var writer = WriteBuffer.PgWriter.Init(DatabaseInfo).WithFlushMode(async ? FlushMode.NonBlocking : FlushMode.Blocking);
```

Comment:
@NinoFloris the EFCore.PG uncovered a bug here. Basically if we reached this point with WriteSpaceLeft equal to zero, Init calls Ensure, which calls SetBuffer (since _buffer is null), which then calls GetMemory which throws OutOfMemoryException. I'm working around it for now by adding a check for WriteSpaceLeft=0 and fluishing just before.

Reply:
Very happy that was found. Leaving the Ensure call out entirely should be an adequate fix. Any subsequent writes should be preceded by a ShouldFlush+Flush anyway.

Reply:
#14

Reply:
Alright it's under test and fixed.